types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,630 @@
+	wl_list_init(&connector->resources);
+	wl_list_insert(&device->connectors, &connector->link);
+
+	struct wl_resource *resource;
+	wl_resource_for_each(resource, &device->resources) {
+		drm_lease_connector_v1_send_to_client(
+				connector, wl_resource_get_client(resource), resource);
+	}
+
+	return true;
+}
+
+void wlr_drm_lease_manager_withdraw_output(
+	struct wlr_drm_lease_manager *manager, struct wlr_output *output) {
+	assert(manager && output);
+	assert(wlr_output_is_drm(output));
+
+	struct wlr_drm_connector *drm_connector =
+		(struct wlr_drm_connector *)output;
+
+	struct wlr_drm_lease_device_v1 *device = NULL;
+	wl_list_for_each(device, &manager->devices, link) {
+		struct wlr_drm_backend *backend = get_drm_backend_from_backend(
+				device->backend);
+		if (backend == drm_connector->backend) {
+			break;
+		}
+	}
+
+	if (!device) {
+		wlr_log(WLR_ERROR, "No wlr_drm_lease_device_v1 associated with the "
+				"given output");
+		return;
+	}
+
+
+	struct wlr_drm_lease_connector_v1 *connector = NULL, *_connector;
+	wl_list_for_each(_connector, &device->connectors, link) {
+		if (_connector->output == output) {
+			connector = _connector;
+			break;
+		}
+	}
+	if (!connector) {
+		return;
+	}
+	assert(connector->active_lease == NULL && "Cannot withdraw a leased output");
+
+	struct wl_resource *wl_resource, *temp;
+	wl_resource_for_each_safe(wl_resource, temp, &connector->resources) {
+		wp_drm_lease_connector_v1_send_withdrawn(wl_resource);
+		wl_resource_set_user_data(wl_resource, NULL);
+		wl_list_remove(wl_resource_get_link(wl_resource));
+		wl_list_init(wl_resource_get_link(wl_resource));
+	}
+
+	wl_resource_for_each(wl_resource, &device->requests) {
+		struct wlr_drm_lease_request_v1 *request =
+			wlr_drm_lease_request_v1_from_resource(wl_resource);
+		request->invalid = true;
+	}
+
+	wl_list_remove(&connector->link);
+	wl_list_init(&connector->link);
+	free(connector);
+}
+
+static void handle_backend_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_drm_lease_device_v1 *device =
+		wl_container_of(listener, device, backend_destroy);
+	struct wl_resource *resource;
+	struct wl_resource *tmp_resource;
+	wl_resource_for_each_safe(resource, tmp_resource, &device->resources) {
+		wl_resource_destroy(resource);
+	}
+
+	wl_resource_for_each_safe(resource, tmp_resource, &device->requests) {
+		wl_resource_destroy(resource);
+	}
+
+	wl_resource_for_each_safe(resource, tmp_resource, &device->leases) {
+		struct wlr_drm_lease_v1 *lease =
+			wlr_drm_lease_v1_from_resource(resource);
+		wlr_drm_lease_v1_revoke(lease);
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector, *tmp_connector;
+	wl_list_for_each_safe(connector, tmp_connector,
+			&device->connectors, link) {
+		wl_list_remove(&connector->link);
+		wl_list_init(&connector->link);
+		free(connector);
+	}
+
+	free(device);
+}
+
+struct wlr_drm_lease_device_v1 *drm_lease_device_v1_create(
+		struct wl_display *display, struct wlr_backend *backend) {
```

Comment:
We should make sure the backend is a DRM backend here.

Reply:
I made it public, and the two places that call it verify that the backend is a drm backend before.